src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3348,25 +3481,27 @@ public static decimal Average<TSource>(this IEnumerable<TSource> source, Func<TS
     /// <summary>
     /// An iterator that can produce an array through an optimized path.
     /// </summary>
-    internal interface IArrayProvider<TElement>
+    internal interface IIListProvider<TElement>
     {
         /// <summary>
         /// Produce an array of the sequence through an optimized path.
         /// </summary>
         /// <returns>The array.</returns>
         TElement[] ToArray();
-    }
 
-    /// <summary>
-    /// An iterator that can produce a <see cref="List{TElement}"/> through an optimized path.
-    /// </summary>
-    internal interface IListProvider<TElement>
-    {
         /// <summary>
         /// Produce a <see cref="List{TElement}"/> of the sequence through an optimized path.
         /// </summary>
         /// <returns>The <see cref="List{TElement}"/>.</returns>
         List<TElement> ToList();
+
+        /// <summary>
+        /// Returns the count of elements in the sequence.
+        /// </summary>
+        /// <param name="onlyIfCheap">If true then the count should only be calculated if doing
+        /// so is quick (sure or likely to be constant time), otherwise -1 should be returned.</param>
+        /// <returns>The number of elements.</returns>
+        int GetCount(bool onlyIfCheap);
```

Comment:
Currently, yes. Through the comment trail, first I had this as a `Count` property and @stephentoub pointed out that someone might use it as a false economy in an optimisation that would be worse when that property wasn't cheap. I said that while it wasn't necessary yet, I'd plans for that (I'm holding back on further changes that build on this as the changes will be massive if I put them all in one PR) and brought a property that reported on whether Count was cheap or not. Stephen pointed out that this would mean two interface calls would be used, and agreeing with him I changed to this. As of right now though, it's still only called with `false`.